mesonbuild/build.py

```diff
@@ -506,7 +506,9 @@ def needs_copy(self) -> bool:
         source dir. Which could happen in situations where a generated source
         only exists in some configurations
         """
-        for files in self.sources.values():
+        for path, files in self.sources.items():
+            if path and files:
+                return True
```

Comment:
This is wrong. The whole point of this is that we only want to actually copy if the structure doesn't already exist in tree. So with this change if I have a structure like:
````
/foo.rs
//lib/mod.rs
//lib/helper.rs
```
and
```meson
s = structured_sources(
  'foo.rs',
  {
    'lib': ['mod.rs', 'helper.rs'],
  }
)
```
Then we could copy even though there's no reason to do so. We should only do a copy if we have a situation where there area  mixture of `Files` that are built and strings or Files that are not built.